src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.RegionBuilder.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Operations
+{
+    internal sealed partial class ControlFlowGraphBuilder
+    {
+        private class RegionBuilder
+        {
+            public readonly ControlFlowGraph.RegionKind Kind;
+            public RegionBuilder Enclosing { get; private set; } = null;
+            public readonly ITypeSymbol ExceptionType;
+            public BasicBlock FirstBlock = null;
+            public BasicBlock LastBlock = null;
+            public ArrayBuilder<RegionBuilder> Regions = null;
+            public ImmutableArray<ILocalSymbol> Locals;
+#if DEBUG
+            private bool _aboutToFree = false;
+#endif 
+
+            public RegionBuilder(ControlFlowGraph.RegionKind kind, ITypeSymbol exceptionType = null, ImmutableArray<ILocalSymbol> locals = default)
+            {
+                Kind = kind;
+                ExceptionType = exceptionType;
+                Locals = locals.NullToEmpty();
+            }
+
+            public bool IsEmpty => FirstBlock == null;
+            public bool HasRegions => Regions?.Count > 0;
+
+#if DEBUG
+            public void AboutToFree() => _aboutToFree = true;
+#endif 
+
+            public void Add(RegionBuilder region)
+            {
+                if (Regions == null)
+                {
+                    Regions = ArrayBuilder<RegionBuilder>.GetInstance();
+                }
+
+#if DEBUG
+                Debug.Assert(region.Enclosing == null || (region.Enclosing._aboutToFree && region.Enclosing.Enclosing == this));
```

Comment:
Don't need the `#if DEBUG` wrapping this.

Reply:
Ah, nevermind, didn't notice the _aboutToFreeUsage.

---
In reply to: [168001615](https://github.com/dotnet/roslyn/pull/24770#discussion_r168001615) [](ancestors = 168001615)

Reply:
> Don't need the #if DEBUG wrapping this.

It is needed because ```_aboutToFree``` available only in debug build. Release is still going to try to bind this expression

---
In reply to: [168001615](https://github.com/dotnet/roslyn/pull/24770#discussion_r168001615) [](ancestors = 168001615)